### PR TITLE
Improve closure calls in the runtime

### LINKED ISSUE #1653: Add specialised versions of CALL_CLOSURES in the runtime
Add versions of CALL_CLOSURES specialised to 1-3 arguments. This will result in more efficient closure calls when arity couldn't be specialised but the number of supplied arguments is small. See also #1545.

This is also relevant for the experimental zkLLVM backend.

### PR DESCRIPTION
**Pull request by lukaszcz**
* Closes #1653 

The running time improvement is 20-30% on benchmarks dominated by closure calls.

### CODE REVIEW

File: runtime/src/juvix/funcall/apply.h [hunk 1]
```diff
@@ -0,0 +1,228 @@
+#ifndef JUVIX_FUNCALL_APPLY_H
+#define JUVIX_FUNCALL_APPLY_H
+
+#include <juvix/funcall/funcall.h>
+
+#define DECL_APPLY_CLOSURE
```
[the remark this answers is not in the document]
Comment (lukaszcz) [reply runtime/src/juvix/funcall/apply.h:6]: Well, yes, it's more like future-proofing. You see below `juvix_apply_closure` is defined as `juvix_ccl_closure`, but one could declare this as a variable separately. Then one only needs to change `DECL_APPLY_CLOSURE`.

[the remark this answers is not in the document]
Comment (paulcadman) [reply runtime/src/juvix/funcall/apply.h:6]: I see, thanks!

**Comment by janmasrovira** [comment]
Just out of curiosity, did you run any benchmark to measure the speedup?

**Comment by lukaszcz** [comment]
No, I didn't bother yet. Though it's interesting how much it actually gives - I will perhaps try tomorrow.

It definitely gives something, especially for closure-heavy programs, and it's very unlikely to ever make things worse (because it's essentially a specialisation of the code we had before to 1-3 arguments).

**Review by paulcadman — APPROVED** [review]

**Comment by lukaszcz** [comment]
The running time improvement is 20-30% on the benchmarks with many closure calls (ackermann, mapfun, cps). It cannot have much effect on benchmarks which don't do any closure calls, because then nothing changes except the C code getting slightly bigger due to the extra specialised apply functions.

**Pull request merged**